### PR TITLE
Invoke "gcc" to build assembler files

### PR DESCRIPTION
**Pull request by jld01**
Enables pre-processing of *.S and *.sx source files.

Part of #666

### CODE REVIEW

**Comment by jld01** [comment]
I am hopeful that this patch will address concerns about the migration of existing projects when invoking `gcc` rather than `as` to build assembly files in managed build projects. My personal opinion is that a switch to using `gcc` is justified here because the current behaviour is not aligned with user expectations for building `*.S` and `*.sx` files.

I have added a command generator for the _Assembler flags_ MBS option that looks for cases where the tool command is `gcc` but the assembler flags option does _not_ include the `-c` flag. In such cases:

- The `-c` flag is added
- Other assembler flags that are not automatically passed to the assembler by the `gcc` wrapper are prefixed with `-Wa,`

For example, a project that was previously building assember files using `as -K -g` would now use `gcc -c -Wa,-K -g`.

**Comment by ruspl-afed** [comment]
Looks good for the simple case. And this covers part of the issues raised during the CDT discussion.

And what about [so nice] MBS inheritance? There are three values changes that can participate to the inheritance chain:
* Tool "command"
* Tool "commandGenerator"
* Option "defaultValue"

How do we ensure that this fix will not break downstream integrations?

**Comment by jld01** [comment]
Thank you the feedback, @ruspl-afed.

> And what about [so nice] MBS inheritance? There are three value changes that can participate to the inheritance chain:
> 
>     * Tool "command"
>     * Tool "commandGenerator"
>     * Option "defaultValue"
> 
> How do we ensure that this fix will not break downstream integrations?

The latest version of this PR accommodates cases where the tool command has been overridden XOR the option default value has been overridden. There is no issue when both have been overridden. There is no issue when neither has been overridden.

This work does _not_ accommodate the possibility that the tool command generator has been overridden while the tool command and option default value have not been overridden. The risk here seems very low.

File: build/org.eclipse.cdt.managedbuilder.gnu.ui/src/org/eclipse/cdt/managedbuilder/gnu/ui/GnuAsmFlagsCommandGenerator.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	private static final Pattern DO_NOT_LINK_PATTERN = Pattern.compile("(^|\\s)-c($|\\s)"); //$NON-NLS-1$
+	private static final Pattern ASM_FLAG_PATTERN = Pattern.compile("-[aDKLR]\\S*"); //$NON-NLS-1$
+
+	@Override
+	public String generateCommand(IOption option, IVariableSubstitutor macroSubstitutor) {
+		String toolCommand = getToolCommand(option, macroSubstitutor);
+		try {
+			if (null != toolCommand && IOption.STRING == option.getValueType() && option.getCommand().isEmpty()) {
+				String optionValue = option.getStringValue();
+				if (toolCommand.equals("gcc")) { //$NON-NLS-1$
```
Comment (Torbjorn-Svensson) [inline build/org.eclipse.cdt.managedbuilder.gnu.ui/src/org/eclipse/cdt/managedbuilder/gnu/ui/GnuAsmFlagsCommandGenerator.java:52]: Sorry to be late the party here, but should this line not match `arm-non-eabi-gcc` for example, or `gcc.exe`? Or is it enough to have a strict check for `gcc`?

Comment (jld01) [reply build/org.eclipse.cdt.managedbuilder.gnu.ui/src/org/eclipse/cdt/managedbuilder/gnu/ui/GnuAsmFlagsCommandGenerator.java:52]: We are checking only whether the tool command has been overridden. We don't care if it has been overridden by the user or by an extender. If the tool command has been overridden, then our new default value (`gcc`) will have no effect on the build and we do not need to modify the assembler flags.

**Pull request merged**